Main Menu Redesign, Part 4

Noordfrees opened the pull request:
Instead of multiple modal windows, we now have 1 non-modal singleton window that changes its content dynamically. Some redesigns to the LaunchMPG screen (got rid of the little map/save selection window).

The various game-running-related codepaths were moved away from WLApplication to those widgets that actually perform the tasks (e.g. a new game is now started in the LaunchSPG).

New handling for emergency saves.

Fixed some memory leaks in the random map generator, and some of the leaks that occur after an emergency save.

This is the last big change in the series. It could not be split into smaller parts unfortunately. I'm planning to have a few more minor changes, e.g. to move the progressscreen for the random match menu into the window, which will need to wait for #4319 to avoid merge conflicts.

Noordfrees committed:
Create new class MenuCapsule

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into main-menu-redesign-4

Noordfrees committed:
Convert SPG workflow

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into main-menu-redesign-4

Noordfrees committed:
Fix emergency save handling

Noordfrees committed:
More emergency saving fixes

Noordfrees committed:
Tutorials

Noordfrees committed:
Campaigns & Scenarios

Noordfrees committed:
Make About window Unique and fix options glitch

Noordfrees committed:
Loadgame and replays

Noordfrees committed:
Fix LAN hosting

Noordfrees committed:
Fix LAN gaming for host and clients

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into main-menu-redesign-4

Noordfrees committed:
Fix lobby

Noordfrees committed:
Login fixes

Noordfrees committed:
Cleanup & emergency save fixes

Noordfrees committed:
fix lobby chat layout

Noordfrees committed:
Convert random game launching

Noordfrees committed:
Beautify MPG map/save selection

Noordfrees committed:
29 files were automatically formatted.

frankystone commented:
This is a bit annoying regarding the 'Back' buttons, but this is similar in master. In general i would expect clicking 'Back' moves me to the previous screen, but this isn't true for some circumstances: 

1. Singleplayer: Hit 'new game' (first screen) -> choose map (second screen) -> Hitting 'Back' now closes the window and the player finds himself in the main menu. I would expect the new game screen is shown.
2. Start Multiplayer lan game (first screen) -> Hit 'Host a new game' (second screen) -> choose map (third screen) -> hit 'Back'  -> the player finds himself in the first screen and the previously chosen map is shown in the list of open games. I would expect the 'Host a new game' screen is shown.

Maybe this is right time to get rid of the 'Map' button (in the upper right) and integrate this in the buttons at the bottom. Instead naming the back-button 'Back' the name of this button can be 'Close' and closes the window. This would be more clear for the user. Not sure if this works well with the multiplayer setup screen, because i am not very familiar with it.

Another thing in Multiplayer: Choose Lan game (first screen) -> 'Host a new Game' (second screen) -> In the small window click 'Cancel'
1. I would expect to get back to the first screen
2. If you click 'Back' then the first screen is shown and in the list of games your entry is shown, but with no map.

The title of the MP Lan game window is also misleading: 'Begin a LAN Game'. Maybe better is 'Lan Game Setup' or similar?

Noordfrees commented:
Um… which version are you using?

Especially this:
> Another thing in Multiplayer: Choose Lan game (first screen) -> 'Host a new Game' (second screen) -> In the small window click 'Cancel'

→ The small window no longer exists in this branch.

Regarding the SPG/choosemap annoyance, this is something I definitely want to address as well, but this should go in a follow-up branch as the behaviour is no worse here than in master

frankystone commented:
Hm, sorry i was at [1aff8cd@main-menu-redesign-4] (Debug), forgot to pull...

frankystone commented:
Rev.: [9cacad8@main-menu-redesign-4] (Debug)

Hm, now it is really annoying in multiplayer: When clicking 'Host a new game', the mouse pointer moves to the button of 'Select Map' and this button is obviously clicked automatically. In result the Dropdown is opened and the tooltip of the button is in front of the first dropdown entry.
![host_new_game](https://user-images.githubusercontent.com/6730556/99179836-82651a00-2721-11eb-9dc2-5d781aacbdeb.jpg)

The solution with the small window was much clearer for me (good that i forgot to pull :-D ). I wouldn't expect the ui does any clicks automatically.

Actually i don't know how it is in master:

When Clicking 'New random map' in the main menu, the background changes like it is when a game is loading ( photography is shown). First i thought that a game is directly startet, but after a short time the Setup screen for Random games comes up and the background changes again and shows a widelands game. I think the players get familiar with this behavior. Nevertheless confusing...

Noordfrees commented:
Well, in master the game also does one click for you – it clicks the Choose Map Or Savegame button immediately to open the little window with the three buttons. Here it clicks the dropdown immediately. 
But if this is annoying or unexpected, I think I'll adopt your suggestion from above then and move the choose map/save button(s) below the Choose Starting Positions checkbox where they'll be always accessible. The right column is big enough for this.

> When Clicking 'New random map' in the main menu, the background changes like it is when a game is loading ( photography is shown). First i thought that a game is directly startet, but after a short time the Setup screen for Random games comes up and the background changes again and shows a widelands game. I think the players get familiar with this behavior. Nevertheless confusing...

This feature has been around for only a few weeks, it was not in b21 yet. Since the majority of players will skip from b21 to 1.0 it does not really matter what we experiment with between versions in terms of getting familiar with something. Anyway I find the current behaviour very annoying and am planning to make it much more intuitive … soonish.

Noordfrees committed:
Move Choose Map/Savegame buttons to bottom of right column

Noordfrees committed:
Cleanup includes

Noordfrees committed:
'src/ui_fsmenu/launch_game.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into main-menu-redesign-4

Noordfrees committed:
'src/wui/minimap.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into main-menu-redesign-4

matthiakl commented:
I agree with @frankystone, that the "back" behavior in the single player screen is confusing. You can right click the window or use the red x to close it, so I think "back" should go to the map selection screen. (Then the "Change map" button can be removed?)

Some issues I encountered:

- Clicking the "Close" button in the About window has no effect.
- When selecting New or Random map in the editor dropdown menu, the menu remains open after returning from the editor.
- You can join your own multiplayer game and crash Widelands: 
  1. Create a LAN game 
  2. Go back -> game is listed as "open"
  3. Join the game, Go back -> game is listed as "closed"
  4. Join the game again:
```
terminate called after throwing an instance of 'WLWarning'
  what():  Widelands could not establish a connection to the given address. Either no Widelands server was running at the supposed port or the server shut down as you tried to connect.
```

Noordfrees committed:
Fix dropdowns and About, and networking glitches

Noordfrees committed:
3 files were automatically formatted.

Noordfrees commented:
Fixed the issues with the dropdowns and the about window.

For LAN, a similar issue can be observed in master when you have multiple Widelands instances and keep opening games and going back. I changed it now so it goes to Closed state directly after the host presses Back, and not only later. 
Just removing the game from the list then, when the state is Closed, would be possible but then we have glitches when observing the progress with another client. So now I instead grey out all non-Open games, and instead of crashing the error message is shown when connecting to an invalid client.

The reason why the LaunchSPG screen shows this weird behaviour is that we actually open *first* the game setup screen, and the map select is stacked *on top of it*. It always was like this; in b21 the back button had special handlings to pretend that the setup screen is stacked on top of the map select screen and not the other way around. Swapping the stacking order to make the handling fully intuitive (both UI-wise and code-wise) is something I'll work on as soon as this branch here is in.

Noordfrees committed:
Merge branch 'master' into main-menu-redesign-4

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into main-menu-redesign-4

Noordfrees committed:
Merged master

Noordfrees committed:
'src/ui_fsmenu/addons.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into main-menu-redesign-4

Noordfrees committed:
Merged master

Noordfrees committed:
Merge branch 'master' into main-menu-redesign-4

Noordfrees committed:
Merge branch 'master' into main-menu-redesign-4

Noordfrees committed:
Merge branch 'master' into main-menu-redesign-4

Noordfrees committed:
Merge branch 'master' into main-menu-redesign-4

matthiakl reviewed (commented):
Code LGTM. Some includes could be cleaned up.
The dropdown menu stays open, if a keyboard shortcut is used (e.g. press S then N)

Noordfrees committed:
Cleanup includes, and close dropdowns when pressing shortcut

Noordfrees commented:
Thanks for the review :)
